docs/docsite/rst/dev_guide/developing_program_flow_modules.rst

```diff
@@ -615,7 +615,10 @@ required
 no_log
 """"""
 
-``no_log`` indicates that the value of the argument should not be logged or displayed.
+``no_log`` accepts a boolean, either ``True`` or ``False``, that indicates explicitly whether or not the argument value should be masked in logs and output.
+
+.. note::
+   In the absence of ``no_log``, the argument value will only be masked in logs and output if the argument name appears to indicate that the argument value is a password or passphrase (such as "admin_password"). To ensure that an argument value always appears in logs and output, and to disable the warning about ``no_log``, set ``no_log`` explicitly to ``False``.
```

Comment:
Actually, I don't think this is accurate. The value is not masked if the parameter name contains something like 'password' — only a warning is issued.
```suggestion
   In the absence of ``no_log``, the parameter value will **not** be masked in logs and output. If the parameter name appears to indicate that the argument value is a password or passphrase (such as "admin_password"), a warning will be shown. To disable the warning on parameters that do not contain sensitive information, set ``no_log`` to ``False``.
```

Reply:
I'm not sure if it's that simple. `log_args[param]` does get initialized to `NOT_LOGGING_PASSWORD`, but I haven't finished looking at where that value ends up. Would you happen to know off-hand?

Reply:
It gets passed to the `log()` method, but it doesn't actually mask anything since `remove_values()` is using the values in `self.list_no_log_values`.

This is probably a bug because I bet the intention was to mask those values if they looked like that contained sensitive information.

Reply:
Oh, nevermind. It _does_ actually get masked in the log message, just not in the module invocation.

```
syslog_msg='Invoked with password_length=NOT_LOGGING_PASSWOR...t_elements
```

Reply:
Ahh. That's what I was trying to test. Thanks.

Reply:
So the value is masked in the syslog, but not the Ansible log.

Reply:
You could technically add the value to `no_log_values` on the module object with each invocation of `_log_invocation` (since it's a `set`) and eat the cost of the call, but that would  result in an inconsistency between the resulting messages, I believe. `"NOT_LOGGING_PASSWORD"` in syslog, and `"NO_LOGGING_PARAMETER"` in the output.

Reply:
For the sake of this PR, should we be documenting the intended behavior, or the current behavior, including the bug?

Reply:
I wonder if that's an omission. I will ask around. It's simple to also mask that value in output.

```patch
diff --git lib/ansible/module_utils/basic.py lib/ansible/module_utils/basic.py
index a96ce6aa5d..14a22bdf59 100644
--- lib/ansible/module_utils/basic.py
+++ lib/ansible/module_utils/basic.py
@@ -1948,6 +1948,7 @@ class AnsibleModule(object):
             elif PASSWORD_MATCH.search(param) and arg_opts.get('type', 'str') != 'bool' and not arg_opts.get('choices', False):
                 # skip boolean and enums as they are about 'password' state
                 log_args[param] = 'NOT_LOGGING_PASSWORD'
+                self.no_log_values.add(self.params[param])
                 self.warn('Module did not set no_log for %s' % param)
             else:
                 param_val = self.params[param]

```

Reply:
We discussed this internally, and adding the suspect value to `self.no_log_values` could potentially cause issues by masking a lot of output strings that don't need to be masked.

Reply:
Potentially. If the consensus is that the responsibility placed on module authors also needs to come with some safeguards then I can understand the logic behind it. Should this be documented that way in the documentation then?

Reply:
Yes. It is up to the module author to mark fields as `no_log`.